File: app/schemas/recipe.py
"""
菜谱相关的 Pydantic 模型
"""
from typing import List, Optional
from pydantic import BaseModel, ConfigDict


# 食材模型
class IngredientBase(BaseModel):
    """食材基础模型"""
    name: str
    quantity: Optional[float] = None
    unit: Optional[str] = None
    text_quantity: str
    notes: Optional[str] = None


class IngredientCreate(IngredientBase):
    """创建食材模型"""
    pass


class IngredientResponse(IngredientBase):
    """食材响应模型"""
    id: int
    
    model_config = ConfigDict(from_attributes=True)


# 步骤模型
class StepBase(BaseModel):
    """步骤基础模型"""
    step: int
    description: str


class StepCreate(StepBase):
    """创建步骤模型"""
    pass


class StepResponse(StepBase):
    """步骤响应模型"""
    id: int
    
    model_config = ConfigDict(from_attributes=True)


# 菜谱模型
class RecipeBase(BaseModel):
    """菜谱基础模型"""
    name: str
    description: Optional[str] = None
    source_path: Optional[str] = None
    image_path: Optional[str] = None
    images: List[str] = []
    category: str
    difficulty: int
    tags: List[str] = []
    servings: int
    prep_time_minutes: Optional[int] = None
    cook_time_minutes: Optional[int] = None
    total_time_minutes: Optional[int] = None


class RecipeCreate(RecipeBase):
    """创建菜谱模型"""
    id: str
    ingredients: List[IngredientCreate] = []
    steps: List[StepCreate] = []
    additional_notes: List[str] = []


class RecipeUpdate(BaseModel):
    """更新菜谱模型"""
    name: Optional[str] = None
    description: Optional[str] = None
    source_path: Optional[str] = None
    image_path: Optional[str] = None
    images: Optional[List[str]] = None
    category: Optional[str] = None
    difficulty: Optional[int] = None
    tags: Optional[List[str]] = None
    servings: Optional[int] = None
    prep_time_minutes: Optional[int] = None
    cook_time_minutes: Optional[int] = None
    total_time_minutes: Optional[int] = None
    ingredients: Optional[List[IngredientCreate]] = None
    steps: Optional[List[StepCreate]] = None
    additional_notes: Optional[List[str]] = None


class RecipeResponse(RecipeBase):
    """菜谱响应模型"""
    id: str
    ingredients: List[IngredientResponse] = []
    steps: List[StepResponse] = []
    additional_notes: List[str] = []
    
    model_config = ConfigDict(from_attributes=True)


class RecipeListItem(BaseModel):
    """菜谱列表项模型（简化版）"""
    id: str
    name: str
    description: Optional[str] = None
    category: str
    difficulty: int
    tags: List[str] = []
    image_path: Optional[str] = None
    total_time_minutes: Optional[int] = None
    
    model_config = ConfigDict(from_attributes=True)

File: zwei/internal/models/recipe.go
package models

import (
	"database/sql/driver"
	"errors"

	"github.com/go-json-experiment/json"
)

// StringSlice JSON 存储的字符串数组
type StringSlice []string

func (s StringSlice) Value() (driver.Value, error) {
	if s == nil {
		return "[]", nil
	}
	return json.Marshal(s)
}

func (s *StringSlice) Scan(value interface{}) error {
	if value == nil {
		*s = []string{}
		return nil
	}

	var bytes []byte
	switch v := value.(type) {
	case []byte:
		bytes = v
	case string:
		bytes = []byte(v)
	default:
		return errors.New("无法将值转换为 StringSlice")
	}

	return json.Unmarshal(bytes, s)
}

// Recipe 菜谱表
type Recipe struct {
	ID               uint        `gorm:"primaryKey;autoIncrement;column:_id" json:"-"`
	RecipeID         string      `gorm:"uniqueIndex;not null;column:recipe_id;size:32" json:"id"`
	Name             string      `gorm:"uniqueIndex;not null;size:128" json:"name"`
	Description      *string     `gorm:"type:text" json:"description"`
	Images           StringSlice `gorm:"type:json;default:'[]'" json:"images"`
	Category         string      `gorm:"index" json:"category"`
	Difficulty       int         `json:"difficulty"`
	Servings         int         `json:"servings"`
	PrepTimeMinutes  *int        `gorm:"column:prep_time_minutes" json:"prep_time_minutes"`
	CookTimeMinutes  *int        `gorm:"column:cook_time_minutes" json:"cook_time_minutes"`
	TotalTimeMinutes *int        `gorm:"column:total_time_minutes" json:"total_time_minutes"`

	// 关联关系（无外键约束，在应用层处理，使用 recipe_id 作为关联字段）
	Ingredients     []Ingredient     `gorm:"foreignKey:RecipeID;references:RecipeID" json:"ingredients"`
	Steps           []Step           `gorm:"foreignKey:RecipeID;references:RecipeID" json:"steps"`
	AdditionalNotes []AdditionalNote `gorm:"foreignKey:RecipeID;references:RecipeID" json:"additional_notes"`
	Tags            []Tag            `gorm:"-" json:"tags"` // 不自动 JOIN，在 service 层手动填充
}

// GetImagePath 获取主图路径（images 数组第一张）
func (r *Recipe) GetImagePath() *string {
	if len(r.Images) > 0 {
		return &r.Images[0]
	}
	return nil
}

func (Recipe) TableName() string {
	return "t_recipe"
}

// Ingredient 食材表
type Ingredient struct {
	ID           uint     `gorm:"primaryKey;autoIncrement;column:_id" json:"-"`
	RecipeID     string   `gorm:"not null;index;column:recipe_id" json:"-"`
	Name         string   `gorm:"not null" json:"name"`
	Category     *string  `gorm:"index" json:"category"`
	Quantity     *float64 `json:"quantity"`
	Unit         *string  `json:"unit"`
	TextQuantity string   `gorm:"not null;column:text_quantity" json:"text_quantity"`
	Notes        *string  `json:"notes"`
}

func (Ingredient) TableName() string {
	return "t_ingredient"
}

// Step 步骤表
type Step struct {
	ID          uint   `gorm:"primaryKey;autoIncrement;column:_id" json:"-"`
	RecipeID    string `gorm:"not null;index;column:recipe_id" json:"-"`
	Step        int    `gorm:"not null" json:"step"`
	Description string `gorm:"not null;type:text" json:"description"`
}

func (Step) TableName() string {
	return "t_step"
}

// AdditionalNote 小贴士表
type AdditionalNote struct {
	ID       uint   `gorm:"primaryKey;autoIncrement;column:_id" json:"-"`
	RecipeID string `gorm:"not null;index;column:recipe_id" json:"-"`
	Note     string `gorm:"not null;type:text" json:"note"`
}

func (AdditionalNote) TableName() string {
	return "t_additional_note"
}
